Apps/frontend/src/views/Oscilloscope.svelte

```diff
@@ -15,73 +15,100 @@
 
   let isEnabled = false;
 
+  // ----- Svelte lifecycle hooks -----
   onMount(() => {
-    socket = new WebSocket(import.meta.env.VITE_BACKEND_WEBSOCKET);
+    socket = createSocket();
+
+    socket.onopen = socketOnOpen;
+    socket.onmessage = socketOnMessage;
+    socket.onclose = socketOnClose;
+  });
+
+  onDestroy(() => {
+    socket.close();
+  });
+
+  // -----business logic functions -----
+  const createSocket = function () {
+    let socket = new WebSocket(import.meta.env.VITE_BACKEND_WEBSOCKET);
     socket.binaryType = "arraybuffer";
 
-    socket.onopen = () => {
-      console.log("Socket opened");
-    };
+    return socket;
+  };
+
+  const socketOnOpen = function (event) {
+    console.log("Socket opened");
+  };
+
+  const socketOnMessage = function (messageEvent) {
+    let samples = new Float64Array(messageEvent.data);
+    if (!isEnabled) return;
+    waveElement.updateBuffer(samples);
+    indicatorElement.update(samples);
+  };
 
-    socket.onclose = function (event) {
-      var reason;
-      // See https://www.rfc-editor.org/rfc/rfc6455#section-7.4.1
-      if (event.code == 1000)
+  const socketOnClose = function (closeEvent) {
```

Comment:
One other thing, can we export this to another file, it clutters the file alot and has no specific logic that warrants it to be in here.